tests/line.c

```diff
@@ -150,70 +150,277 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
     SVPNG_END(); /* IEND chunk {} */
 }
 
-#include <math.h>  // ceilf(), floorf(), fminf(), fmaxf(), sinf(), cosf(), sqrtf()
 #include <string.h>  // memset()
+#define Q (20)
+#define Q_PI (1686629713 >> (29 - Q))
+
+/* 32-bit Q format */
+typedef int32_t q_fmt;
+
+/* 64-bit Q format buffer */
+typedef int64_t q_buf;
+
+/* format convertion: float to Q format */
+#define f2Q(x) ((q_fmt) ((x) * (1 << Q)))
+
+/* format convertion: Q format to float */
+#define Q2f(x) (((float) (x)) / (1 << Q))
+
+/* format convertion: Q format to int */
+#define Q2I(x) ((int) ((x) >> Q))
+
+/* format convertion: int to Q format */
+#define I2Q(x) ((q_fmt) ((x) << Q))
+
+/* maximum value of Q format */
+#define QFMT_MAX 0x7FFFFFFF
+
+/* minimum value of Q format */
+#define QFMT_MIN 0x80000000
+
+/* addition of Q format value */
+static inline q_fmt q_add(q_fmt a, q_fmt b)
+{
+    q_buf tmp = (q_buf) a + (q_buf) b;
+    if (tmp > (q_buf) QFMT_MAX)
+        return (q_fmt) QFMT_MAX;
+    if (tmp * -1 >= (q_buf) QFMT_MIN)
+        return (q_fmt) QFMT_MIN;
+    return (q_fmt) tmp;
+};
+
+/* multiplication of Q format value */
+static inline q_fmt q_mul(q_fmt a, q_fmt b)
+{
+    q_buf tmp = (q_buf) a * (q_buf) b;
+
+    /* rounding and resize */
+    tmp += (q_buf) (1 << (Q - 1));
+    tmp >>= Q;
+
+    /* check overflow */
+    if (tmp > (q_buf) QFMT_MAX)
+        return (q_fmt) QFMT_MAX;
+    if (tmp * -1 >= (q_buf) QFMT_MIN)
+        return (q_fmt) QFMT_MIN;
+    return (q_fmt) tmp;
+}
+
+/* division of Q format value */
+static inline q_fmt q_div(q_fmt a, q_fmt b)
+{
+    q_buf tmp = (q_buf) a << Q;
+    if ((tmp >= 0 && b >= 0) || (tmp < 0 && b < 0)) {
+        tmp += (b >> 1);
+    } else {
+        tmp -= (b >> 1);
+    }
+    return (q_fmt) (tmp / b);
+}
+
+/* return the largest integral value that is not greater than x */
+static inline q_fmt floorq(q_fmt x)
```

Comment:
Rename to `qfloor`. Change the corresponding functions, ensuring prefix `q`.

Reply:
Should I also rename basic arithmetic functions?  e.g. q_add() rename to qadd().

Reply:
No, let's unify with the consistent `q_` prefix.